complete SigHash enum and make updates to stop using ordinal

mruddy opened the pull request:
Complete the SigHash enum and make updates to stop using the enum's ordinal while preserving the original ordinal value for any existing code that might use it.

In my own code, I found that I was using an enum defined like below. The changes in this PR, deviate slightly from my original code in order to preserve the ordinal values of the original bitcoinj enum should any external code depend on it still. All in all, I think these changes make the SigHash enum more complete and make the code a little easier to read (less brittle with the removal of the ordinal dependency in the core code) with fewer casts. Let me know if I missed something, but I think my references checks were pretty complete and all the bitcoinj projects still compile and pass tests.

```
public enum Sighash {
  UNSET(0), ALL(1), NONE(2), SINGLE(3), ANYONECANPAY(0x80), ANYONECANPAY_ALL(0x81), ANYONECANPAY_NONE(0x82), ANYONECANPAY_SINGLE(0x83);
  private final int type;

  /**
   * @param type
   */
  private Sighash(final int type) {
    this.type = type;
  }

  /**
   * @return the type as a byte
   */
  public byte byteValue() {
    return (byte) this.type;
  }

  /**
   * @return the type
   */
  public int intValue() {
    return this.type;
  }
}
```

schildbach commented:
I'm fine with embedding the value into the enum. I'm not so sure about the new enum values, isn't ANYONECANPAY more like a flag? Finally, I'd make `private final int type;` public, so the `intValue()` method isn't needed.

schildbach commented on core/src/main/java/org/bitcoinj/core/Transaction.java:
```diff
@@ -466,9 +466,37 @@ public void setUpdateTime(Date updatedAt) {
+         * @param type
+         */
+        private SigHash(final int type) {
+          this.type = type;
```
Mind the spacing here. 4 spaces per level.

schildbach commented on core/src/main/java/org/bitcoinj/core/Transaction.java:
```diff
@@ -466,9 +466,37 @@ public void setUpdateTime(Date updatedAt) {
+        ANYONECANPAY_ALL(0x81),
+        ANYONECANPAY_NONE(0x82),
+        ANYONECANPAY_SINGLE(0x83),
+        UNSET(0);
```
Do we need an UNSET value? In the Java world, this would typically be represented as `null`.

mruddy commented:
@schildbach OK, updated. I made the type publicly accessible. I had it private and used intValue as a stylistic thing to parallel how things are done in the java.lang wrapper classes. But, no big deal, I changed it.

About the enum values:
ANYONECANPAY, SINGLE, and NONE are all really treated like flags, reference https://github.com/bitcoin/bitcoin/blob/a38107643f3f01cd92a4e6da9e7a9d025770ff37/src/script/interpreter.cpp#L1016-L1105

ALL is default behavior which is why 0x00 and 0x80 are valid also.

0x00 is equivalent to SIGHASH_ALL and 0x80 is like SIGHASH_ANYONECANPAY|SIGHASH_ALL.

Using 0x00 and 0x80 are treated as non-standard, but I don't see a consensus rule prohibiting their use.
In fact, that's why this transaction mentioned in a comment in the bitcoinj code is valid (it uses a 0x00 sighash type): https://blockchain.info/tx/c99c49da4c38af669dea436d3e73780dfdb6c1ecf9958baa52960e8baee30e73?show_adv=true

That's why I added UNSET as an enum value. I didn't have a better name. With this explanation though, it might make more sense why I named it that.

I updated some comments too. It's all fundamentally pretty much the same as the previous commit though.

schildbach commented on core/src/test/java/org/bitcoinj/core/FullBlockTestGenerator.java:
```diff
@@ -759,7 +759,7 @@ public boolean add(Rule element) {
                     try {
                         ByteArrayOutputStream bos = new UnsafeByteArrayOutputStream(73);
                         bos.write(coinbaseOutKey.sign(hash).encodeToDER());
-                        bos.write(SigHash.SINGLE.ordinal() + 1);
```
Really byteValue() here? ordinal()+1 was an int, no?

mruddy replied:
yep, definitely. both have the same result in this case. but, not using byteValue() avoids some unnecessary casting so i'll update this in my next commit. i only used byteValue() because i liked the readability.

schildbach commented on core/src/test/java/org/bitcoinj/core/FullBlockTestGenerator.java:
```diff
@@ -830,7 +830,7 @@ public boolean add(Rule element) {
                             ByteArrayOutputStream bos = new UnsafeByteArrayOutputStream(
                                     73);
                             bos.write(coinbaseOutKey.sign(hash).encodeToDER());
-                            bos.write(SigHash.SINGLE.ordinal() + 1);
```
And here?

schildbach commented on core/src/main/java/org/bitcoinj/core/Transaction.java:
```diff
@@ -460,16 +460,34 @@ public void setUpdateTime(Date updatedAt) {
+        }
     }
+
     public static final byte SIGHASH_ANYONECANPAY_VALUE = (byte) 0x80;
```
Is this still needed? One could write ANYONECANPAY.type too.

mruddy replied:
sure, i'll update references to that in my next commit and get rid of it.

mruddy replied:
on second thought, i remember that left that around because it was a public member that existing code could depend on. i think i'll leave it as a public member, but remove references to it. it effectively becomes deprecated. i'll probably add a comment that it's now deprecated and to use the enum.

schildbach commented on core/src/main/java/org/bitcoinj/core/Transaction.java:
```diff
@@ -460,16 +460,34 @@ public void setUpdateTime(Date updatedAt) {
+        ANYONECANPAY_SINGLE(0x83),
+        UNSET(0); // Caution: Using this type in isolation is non-standard. Treated similar to ALL.
+
+        public final int type;
```
Maybe rename this to "value"?

mruddy replied:
sure, will do in my next commit. the same thing crossed my mind before. i left it named "type" because people sometimes refer to the "sighash type". but "value" makes sense when reading the code that uses it.

schildbach commented:
Thanks for the clarification. It feels weird an enum as a bitfield, but you're right it was used like that already before.

schildbach commented:
Looks good! I merged this. Thanks very much!
